File: api/src/main/java/io/perryquotes/api/quote/author/AuthorService.java
package io.perryquotes.api.quote.author;

import io.perryquotes.api.base.BaseEntityService;
import io.perryquotes.api.error.EntityNotFoundException;
import io.perryquotes.api.error.InvalidDataException;
import org.slf4j.Logger;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.annotation.Validated;

import javax.validation.Valid;
import java.util.Optional;
import java.util.UUID;

import static java.lang.String.format;

@Service
@Validated
public class AuthorService extends BaseEntityService<Author> {

  private final AuthorRepository repository;

  public AuthorService(final Logger log, final AuthorRepository repository) {
    super(log);
    this.repository = repository;
  }

  @Transactional(readOnly = true)
  public Optional<Author> findByName(String name) {
    return repository.findByName(name);
  }

  @Transactional
  public Author create(@Valid final AuthorRecord request) {
    var name = request.name();
    if (repository.findByName(name).isPresent()) {
      throw new InvalidDataException(format("Author with name %s already exists", name));
    }
    var created = repository.save(new Author(name));
    log.debug(format("Created Author: %s", created));
    return created;
  }

  @Transactional
  public Author update(UUID uuid, @Valid final AuthorRecord authorRecord) {
    var updated = repository.findByUuid(uuid)
      .map(author -> repository.save(author.setName(authorRecord.name())))
      .orElseThrow(() -> new EntityNotFoundException(Author.class, "uuid", uuid.toString()));
    log.debug(format("Author %s updated: %s", uuid, updated));
    return updated;
  }

  @Override
  protected AuthorRepository getRepository() {
    return repository;
  }
}

File: core/src/main/kotlin/io/perryquotes/core/author/AuthorService.kt
package io.perryquotes.core.author

import io.perryquotes.core.exception.InvalidDataException
import io.perryquotes.core.jooq.tables.pojos.AuthorEntity
import jakarta.validation.Valid
import mu.KLogging
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.validation.annotation.Validated
import java.util.*

@Service
@Validated
class AuthorService(private val repository: AuthorRepository) {

    companion object : KLogging()

    @Transactional(readOnly = true)
    fun getByUuid(uuid: UUID): AuthorModel? {
        logger.debug { "AuthorService.getByUuid = $uuid" }
        return repository.findByUuid(uuid)?.toModel()
    }

    @Transactional(readOnly = true)
    fun getAll(): List<AuthorModel> {
        logger.debug { "AuthorService.getAll" }
        return repository.findAll().map { it.toModel() }
    }

    @Transactional
    fun create(@Valid request: AuthorRequest): AuthorModel {
        logger.debug { "AuthorService.create = $request" }
        if (repository.findByName(request.name) != null) {
            throw InvalidDataException("Author[name=${request.name}] already exists")
        }
        val created = repository.create(AuthorEntity(name = request.name))
        logger.info { "Created Author: $created" }
        return created.toModel()
    }

    @Transactional
    fun update(uuid: UUID, @Valid request: AuthorRequest): AuthorModel {
        logger.debug { "AuthorService.update = $uuid, $request" }
        if (repository.findByName(request.name) != null) {
            throw InvalidDataException("Author[name=${request.name}] already exists")
        }
        val updated = repository.findByUuid(uuid)?.let {
            repository.update(it.copy(name = request.name))
        } ?: throw authorNotFound(uuid)
        logger.info { "Updated Author[uuid=$uuid]: $updated" }
        return updated.toModel()
    }

    @Transactional
    fun delete(uuid: UUID): Boolean {
        logger.debug { "AuthorService.delete = $uuid" }
        return repository.findByUuid(uuid)?.let { repository.delete(uuid) == 1
        }  ?: throw authorNotFound(uuid)
    }

    private fun authorNotFound(uuid: UUID): AuthorNotFoundException =
        AuthorNotFoundException("Author[uuid=$uuid] not found")


}
